Use the name provided on CLI to create an Operator backed service

dharmit opened the pull request:
**What type of PR is this?**
/kind feature
/kind code-refactoring

**What does does this PR do / why we need it**:
This PR lets the user create a service named `myservice` when they do something like `odo service create etcdoperator.v0.9.4-clusterwide --crd EtcdCluster myservice`. This wasn't working earlier for Operator Hub backed services

It also does some refactoring of the code in `pkg/odo/cli/service/create.go`. Part of the business logic is moved to `service` package.
**Which issue(s) this PR fixes**:

Fixes #3180

**How to test changes / Special notes to the reviewer**:
```sh
$ odo service create etcdoperator.v0.9.4-clusterwide --crd EtcdCluster myservice

$ odo service list
NAME                TYPE                      AGE
myservice            EtcdCluster               4s
```

dharmit commented on pkg/odo/cli/service/create.go:
```diff
@@ -270,43 +282,34 @@ func (o *ServiceCreateOptions) Validate() (err error) {
-					// this is satisfied if user has specified operator but not
-					// a CRD name
-					return errors.New("Please specify service name along with the operator name")
+			if o.ServiceName != "" {
```
This is the main part from where we ensure that user provided name is used to create the service.

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -96,6 +97,8 @@ type ServiceCreateOptions struct {
 	// TODO: remove this after service create's interactive mode supports creating operator backed services
 	fromFile string
+	// For accessing information from Devfile
+	EnvSpecificInfo *envinfo.EnvSpecificInfo
```
Yes we can get it from the context

mik-dass commented on pkg/service/service.go:
```diff
@@ -65,6 +65,71 @@ func CreateService(client *occlient.Client, serviceName string, serviceType stri
+// CheckCRExists checks if the CR provided by the user in the YAML file exists in the namesapce
+// It returns a CR (string representation) and CSV (Operator) upon successfully
+// able to find them, an error otherwise.
+func CheckCRExists(client *kclient.Client, crd map[string]interface{}) (string, olm.ClusterServiceVersion, error) {
```
How about just GetCR() or GetCSV()?

girishramnani commented on pkg/odo/cli/service/create.go:
```diff
@@ -447,56 +450,57 @@ func NewCmdServiceCreate(name, fullName string) *cobra.Command {
-				return csv, nil
+// setServiceName modifies the CRD to contain user provided name on the CLI
+// instead of using the default one in almExample
+func setServiceName(crd interface{}, name string) (map[string]interface{}, error) {
```
I think we need to have a struct called `DynamicCRD` which holds metdata, kind and other such fields with the original `crd` as a `map[string]interface{}`. 
So something like
```
type DynamicCRD struct {
Metadata Metadata
Kind string
RawCRD map[string]interface{}
}
```
and then this struct could have `getServiceNameFromCRD`,` setServiceName`, `CheckCRExists` etc

All the functions added so far seem to indicate that this refactor is gonna help us so I would suggest you please do the refactor as part of this PR 🙂 . And this change is related to most of the code added as well.

If you find any other opportunity to refactor then please do that too as part of this PR. And we shouldn't delay visible refactoring opportunities because else we would never do them.

dharmit replied:
> So something like
> 
> ```
> type DynamicCRD struct {
> Metadata Metadata
> Kind string
> RawCRD map[string]interface{}
> }
> ```
> 
> and then this struct could have `getServiceNameFromCRD`,` setServiceName`, `CheckCRExists`(not sure about if this fits but consider if it does) etc

Makes sense to me. I'll make the change and ping for review after that.

> All the functions added so far seem to indicate that this refactor is gonna help us so I would suggest you please do the refactor as part of this PR :slightly_smiling_face: . And this change is related to most of the code added as well.
> 
> If you find any other opportunity to refactor then please do that too as part of this PR. And we shouldn't delay visible refactoring opportunities because else we would never do them.

We're on the same page here. https://github.com/openshift/odo/pull/3509/#pullrequestreview-444494596

dharmit commented:
@girishramnani I've added the struct as you suggested and based on what seemed relevant to me. I didn't find a strong reason to put `CheckCRExists` in the scope of it. PTAL and let me know!

girishramnani commented:
/approve

dharmit commented:
/hold 

There's work pending on this since I got busy with #3557 #3561 & #3564. Will look at this now.

dharmit commented:
/hold cancel

ready for review!

dharmit commented:
Failure in tests/code unmodified by this PR.
/test v4.5-integration-e2e
```
 • Failure [2.032 seconds]
odo project command tests
/go/src/github.com/openshift/odo/tests/integration/project/cmd_project_test.go:15
  odo machine readable output on empty project
  /go/src/github.com/openshift/odo/tests/integration/project/cmd_project_test.go:52
    should be able to list current project [It]
    /go/src/github.com/openshift/odo/tests/integration/project/cmd_project_test.go:53
    Expected
      
        <string>: {"apiVersion":"odo.dev/v1alpha1","items":[{"apiVersion":"odo.dev/v1alpha1","kind":"Project","metadata":{"creationTimestamp":null,"name":"ci-operator-hub-project","namespace":"ci-operator-hub-project"},"spec":{},"status":{"active":false}},{"apiVersion":"odo.dev/v1alpha1","kind":"Project","metadata":{"creationTimestamp":null,"name":"dclylequeg","namespace":"dclylequeg"},"spec":{},"status":{"active":false}},{"apiVersion":"odo.dev/v1alpha1","kind":"Project","metadata":{"creationTimestamp":null,"name":"mkwebrlckl","namespace":"mkwebrlckl"},"spec":{},"status":{"active":false}},{"apiVersion":"odo.dev/v1alpha1","kind":"Project","metadata":{"creationTimestamp":null,"name":"myproject","namespace":"myproject"},"spec":{},"status":{"active":false}},{"apiVersion":"odo.dev/v1alpha1","kind":"Project","metadata":{"creationTimestamp":null,"name":"pawnpeuwqn","namespace":"pawnpeuwqn"},"spec":{},"status":{"active":false}},{"apiVersion":"odo.dev/v1alpha1","kind":"Project","metadata":{"creationTimestamp":null,"name":"pqyqmqqbcq","namespace":"pqyqmqqbcq"},"spec":{},"status":{"active":false}},{"apiVersion":"odo.dev/v1alpha1","kind":"Project","metadata":{"creationTimestamp":null,"name":"rmjsdy","namespace":"rmjsdy"},"spec":{},"status":{"active":false}}],"kind":"List","metadata":{}}
    to contain substring
        <string>: {"apiVersion":"odo.dev/v1alpha1","kind":"Project","metadata":{"creationTimestamp":null,"name":"teogylikvd","namespace":"teogylikvd"},"spec":{},"status":{"active":true}}
    /go/src/github.com/openshift/odo/tests/integration/project/cmd_project_test.go:59 
```

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -447,56 +477,50 @@ func NewCmdServiceCreate(name, fullName string) *cobra.Command {
+// setServiceName modifies the CRD to contain user provided name on the CLI
+// instead of using the default one in almExample
+func (d *DynamicCRD) setServiceName(name string) error {
+	m := d.OriginalCRD["metadata"].(map[string]interface{})
```
something more descriptive than `m`

mohammedzee1000 replied:
While I understand usage of small variable names in small contexts, this might be a little too small, considering it is also a map

dharmit replied:
Fixed.

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -447,56 +477,50 @@ func NewCmdServiceCreate(name, fullName string) *cobra.Command {
-	return olmv1alpha1.ClusterServiceVersion{}, errors.New("Could not find the requested cluster resource")
 
+	// if we reach this point, it's likely becuase the CRD doesn't have
+	// metadata or doesn't have metadata.name
```
But if it doesn't have `metadata.name`, we are setting it up above, right? Or am I missing something?

mohammedzee1000 replied:
True. I feel it is necessary though as it catches the case where `metadata` is missing. If its there and name is not set, it is set here so comment needs fixing

dharmit replied:
@mohammedzee1000 +1.

Removed this part of the code since `metadata` is surely going to be there in an alm example. If it's the case of doing `--from-file`, we're now validating the yaml before setting name. That should suffice, IMO.

If you folks think I'm missing a case, let me know.

mik-dass commented on pkg/service/service.go:
```diff
@@ -307,6 +348,41 @@ func getGVRFromCR(cr *olm.CRDDescription) (group, version, resource string, err
+	return almExample, nil
+}
+
+func getAlmExample(almExamples []map[string]interface{}, crd, operator string) (map[string]interface{}, error) {
```
some comments

dharmit replied:
Added.

mik-dass commented:
Interactive mode seems to panic
```
[mrinaldas@localhost project]$ odo service create
I0722 16:43:08.028402   21831 cert_rotation.go:137] Starting client certificate rotation controller
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x1496167]

goroutine 1 [running]:
github.com/openshift/odo/pkg/occlient.(*Client).GetClusterServiceClasses(0x0, 0x1b773c0, 0x1, 0x0, 0xc000238f40, 0x40f59d)
        /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient.go:2364 +0x37
github.com/openshift/odo/pkg/occlient.(*Client).GetServiceClassesByCategory(0x0, 0xc0002e2cb0, 0x197, 0x0)
        /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient.go:2562 +0x51
github.com/openshift/odo/pkg/odo/cli/service.(*ServiceCreateOptions).Complete(0xc0007fc1e0, 0xc000052480, 0x6, 0xc0001d6c80, 0x301cd70, 0x0, 0x0, 0xc0000d2f00, 0x2)
        /home/mrinaldas/go/src/github.com/openshift/odo/pkg/odo/cli/service/create.go:142 +0x3cb
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x1f799c0, 0xc0007fc1e0, 0xc0001d6c80, 0x301cd70, 0x0, 0x0)
        /home/mrinaldas/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:29 +0xb8
github.com/openshift/odo/pkg/odo/cli/service.NewCmdServiceCreate.func1(0xc0001d6c80, 0x301cd70, 0x0, 0x0)
```
On master, it just throws a error.

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -109,7 +112,15 @@ func (o *ServiceCreateOptions) Complete(name string, cmd *cobra.Command, args []
+		}
+
+		o.EnvSpecificInfo = envInfo
+		o.Context = genericclioptions.NewDevfileContext(cmd)
```
Since occlient isn't there, it panics when interactive mode is triggered https://github.com/openshift/odo/pull/3509#issuecomment-662395711

mik-dass replied:
I think we can also have an option to provide the name interactively.

dharmit replied:
Interactive mode is going to be a completely different task. We're not looking at it for v2. Mainly because there's not completely clarity about the approach we should take for it.

mik-dass replied:
Ok 👍

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,31 +252,40 @@ func (o *ServiceCreateOptions) Validate() (err error) {
-			return err
-		}
-		if o.CustomResource != "" {
+			o.ServiceName, err = d.getServiceNameFromCRD()
```
shouldn't we use the user provided name in case of creation from a file?

dharmit replied:
User has the option to do that when they provide a file. If they're editing other stuff in the file, expecting them to edit the name is not a big ask, IMO. Plus, the `--from-file` feature is not here to stay. It's gonna go away once we have #2785.

dharmit replied:
Added this to avoid having long discussion. Added tests for it as well.

dharmit commented:
> Interactive mode seems to panic
> 
> ```
> [mrinaldas@localhost project]$ odo service create
> I0722 16:43:08.028402   21831 cert_rotation.go:137] Starting client certificate rotation controller
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x1496167]
> 
> goroutine 1 [running]:
> github.com/openshift/odo/pkg/occlient.(*Client).GetClusterServiceClasses(0x0, 0x1b773c0, 0x1, 0x0, 0xc000238f40, 0x40f59d)
>         /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient.go:2364 +0x37
> github.com/openshift/odo/pkg/occlient.(*Client).GetServiceClassesByCategory(0x0, 0xc0002e2cb0, 0x197, 0x0)
>         /home/mrinaldas/go/src/github.com/openshift/odo/pkg/occlient/occlient.go:2562 +0x51
> github.com/openshift/odo/pkg/odo/cli/service.(*ServiceCreateOptions).Complete(0xc0007fc1e0, 0xc000052480, 0x6, 0xc0001d6c80, 0x301cd70, 0x0, 0x0, 0xc0000d2f00, 0x2)
>         /home/mrinaldas/go/src/github.com/openshift/odo/pkg/odo/cli/service/create.go:142 +0x3cb
> github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x1f799c0, 0xc0007fc1e0, 0xc0001d6c80, 0x301cd70, 0x0, 0x0)
>         /home/mrinaldas/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:29 +0xb8
> github.com/openshift/odo/pkg/odo/cli/service.NewCmdServiceCreate.func1(0xc0001d6c80, 0x301cd70, 0x0, 0x0)
> ```
> 
> On master, it just throws a error.

Is it with experimental mode enabled or disabled? Interactive mode is not expected to work till we have https://github.com/openshift/odo/issues/2799 in.

girishramnani commented:
But it shouldn’t panic as well

mik-dass commented:
> Is it with experimental mode enabled or disabled? Interactive mode is not expected to work till we have #2799 in.

It's with experimental mode enabled. I agree with @girishramnani , it shouldn't panic. Maybe instead of triggering interactive mode, just error out that it is not supported as of now.

dharmit commented:
> > Is it with experimental mode enabled or disabled? Interactive mode is not expected to work till we have #2799 in.
> 
> It's with experimental mode enabled. I agree with @girishramnani , it shouldn't panic. Maybe instead of triggering interactive mode, just error out saying it is not supported as of now.

Agreed. My bad. Fixed. Added test for it.

dharmit commented:
Failure unrelated to the PR:
```

• Failure [31.321 seconds]
odo generic
/go/src/github.com/openshift/odo/tests/integration/generic_test.go:15
  Overwriting build timeout for git component
  /go/src/github.com/openshift/odo/tests/integration/generic_test.go:143
    should pass to build component if the given build timeout is more than the default(300s) value [It]
[odo] I0723 07:06:24.813616   10002 odo.go:72] Could not get the latest release information in time. Never mind, exiting gracefully :)
[odo]  ✓  Deleted project : pkqomjeyfh
Deleting dir: /tmp/194451013

```

/test v4.5-integration-e2e

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,76 +273,99 @@ func (o *ServiceCreateOptions) Validate() (err error) {
+			if o.ServiceName != "" && !o.DryRun {
+				// First check if service with provided name already exists
+				svcFullName := strings.Join([]string{o.CustomResource, o.ServiceName}, "/")
+				exists, _ := svc.OperatorSvcExists(o.KClient, svcFullName)
```
Any reason to ignore the error here?

dharmit replied:
We're checking if the service exists. The function returns `true` if service exists and an error if it fails to find a service with the given name. In `odo service create`, we are concerned with a pre-existing service with the same name. It's not important if it fails to find a service with the given name in `odo service create`. Hence, we're ignoring the error here.

mik-dass replied:
We can use kerrors package to differentiate between the error e.g https://github.com/openshift/odo/blob/c89c1ab1d730fb8495f91ce6ea0ed7ff8f784711/pkg/occlient/occlient.go#L3316-L3320

We need to check the error as in case of a network error, it's better to inform the user and stop the execution. Otherwise it might lead to confusion.

dharmit replied:
I have made the change based on our discussion but, tbh, I'm not so sure about how kerrors works. :man_shrugging:

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,76 +273,99 @@ func (o *ServiceCreateOptions) Validate() (err error) {
+			if o.ServiceName != "" && !o.DryRun {
+				// First check if service with provided name already exists
+				svcFullName := strings.Join([]string{o.CustomResource, o.ServiceName}, "/")
+				exists, _ := svc.OperatorSvcExists(o.KClient, svcFullName)
```
Any reasons to inform the error here too?

dharmit replied:
Same as https://github.com/openshift/odo/pull/3509/#discussion_r459906366.

cdrage replied:
IMO we should never do this :( I agree with @mik-dass that we should check for error regardless

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -447,56 +506,46 @@ func NewCmdServiceCreate(name, fullName string) *cobra.Command {
+	metaMap := d.OriginalCRD["metadata"].(map[string]interface{})
+
+	for k := range metaMap {
+		if k == "name" {
```
Something more descriptive than `k`

dharmit replied:
It actually helps to keep short names at places like these. IMO, it's not always helpful/productive to use long names. `for` loops are one such place. And this one in particular doesn't make sense to me because we're iterating over a `map` which is evident from the name `metaMap`. So it's quite straightforward that if you iterate as `for k, v := range metaMap`, `k` is key and `v` is value, whereas if you iterate `for k := range metaMap`, `k` is the key. I do not think we need a more descriptive name here.

mik-dass replied:
OK

dharmit commented:
```
 {"component":"entrypoint","file":"prow/entrypoint/run.go:242","func":"k8s.io/test-infra/prow/entrypoint.gracefullyTerminate","level":"error","msg":"Process gracefully exited before 30m0s grace period","severity":"error","time":"2020-07-23T17:59:28Z"} 
```
/test v4.5-integration-e2e

dharmit commented on pkg/odo/cli/service/create.go:
```diff
@@ -119,9 +133,22 @@ func (o *ServiceCreateOptions) Complete(name string, cmd *cobra.Command, args []
+
+		// if user has just used "odo service create", simply return
+		if o.fromFile == "" && len(args) == 0 {
+			return
```
We're doing that in `Validate` function which ensures that user doesn't do something like:
- `odo service create`
- `odo service create myoperator` (no `--crd` provided here)

and similar invalid commands.

mik-dass replied:
OK

dharmit commented:
Failure unrelated to the PR:
```
 Setting current dir to: /go/src/github.com/openshift/odo/tests/integration
Deleting dir: /tmp/202213657
• Failure [128.096 seconds]
odo generic
/go/src/github.com/openshift/odo/tests/integration/generic_test.go:15
  creating component with an application and url
  /go/src/github.com/openshift/odo/tests/integration/generic_test.go:118
    should create the component in default application [It]
[Fail] odo generic creating component with an application and url [It] should create the component in default application 
/go/src/github.com/openshift/odo/tests/helper/helper_run.go:34
Ran 179 of 179 Specs in 4351.508 seconds
FAIL! -- 178 Passed | 1 Failed | 0 Pending | 0 Skipped
```
/test v4.5-integration-e2e

cdrage commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,76 +273,99 @@ func (o *ServiceCreateOptions) Validate() (err error) {
+			if o.ServiceName != "" && !o.DryRun {
+				// First check if service with provided name already exists
+				svcFullName := strings.Join([]string{o.CustomResource, o.ServiceName}, "/")
+				exists, _ := svc.OperatorSvcExists(o.KClient, svcFullName)
```
Agreed that we shouldn't ignore the error. Otherwise, rest of the code LGTM.

cdrage commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,76 +273,99 @@ func (o *ServiceCreateOptions) Validate() (err error) {
+				svcFullName := strings.Join([]string{o.CustomResource, o.ServiceName}, "/")
+				exists, _ := svc.OperatorSvcExists(o.KClient, svcFullName)
+				if exists {
+					return fmt.Errorf("Service %q already exists. Please provide a different name or delete the existing service first.", svcFullName)
```
not allowed to have sentence capitalization in errors returning.

cdrage commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,76 +273,99 @@ func (o *ServiceCreateOptions) Validate() (err error) {
+				svcFullName := strings.Join([]string{o.CustomResource, o.ServiceName}, "/")
+				exists, _ := svc.OperatorSvcExists(o.KClient, svcFullName)
+				if exists {
+					return fmt.Errorf("Service %q already exists. Please provide a different name or delete the existing service first.", svcFullName)
```
remove capitalization

cdrage commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,76 +273,99 @@ func (o *ServiceCreateOptions) Validate() (err error) {
+			// correct way is to execute:
+			// `odo service create <operator-name> --crd <crd-name>`
+
+			return fmt.Errorf("Please use a valid command to start an Operator backed service. Desired format: %q", "odo service create <operator-name> --crd <crd-name>")
```
remove capitalization

cdrage commented on pkg/odo/cli/service/create.go:
```diff
@@ -447,56 +506,46 @@ func NewCmdServiceCreate(name, fullName string) *cobra.Command {
+	metadata, ok := d.OriginalCRD["metadata"].(map[string]interface{})
+	if !ok {
+		// this condition is satisfied if there's no metadata at all in the provided CRD
+		return fmt.Errorf("Couldn't find \"metadata\" in the yaml. Need metadata start the service")
```
remove captalization

cdrage commented on pkg/odo/cli/service/create.go:
```diff
@@ -447,56 +506,46 @@ func NewCmdServiceCreate(name, fullName string) *cobra.Command {
+		return nil
 	}
-	return nil, errors.Errorf("Could not find example yaml definition for %q service in %q operator's definition.\nPlease provide a file containing yaml specification to start the service from operator\n", crd, operator)
+	return fmt.Errorf("Couldn't find metadata.name in the yaml. Provide a name for the service")
```
remove captalization

cdrage commented on pkg/service/service.go:
```diff
@@ -66,6 +66,37 @@ func CreateService(client *occlient.Client, serviceName string, serviceType stri
+	csv, err := doesCRExist(cr, csvs)
+	if err != nil {
+		return cr, olm.ClusterServiceVersion{},
+			fmt.Errorf("Could not find specified service/custom resource: %s\nPlease check the \"kind\" field in the yaml (it's case-sensitive)", cr)
```
remove capitalization

cdrage commented on pkg/service/service.go:
```diff
@@ -66,6 +66,37 @@ func CreateService(client *occlient.Client, serviceName string, serviceType stri
+			}
+		}
+	}
+	return olm.ClusterServiceVersion{}, errors.New("Could not find the requested cluster resource")
```
remove capitalization

cdrage commented on pkg/service/service.go:
```diff
@@ -316,9 +343,19 @@ func getGVKRFromCR(cr olm.CRDDescription) (group, version, kind, resource string
+			return GetGVRFromCR(&custRes)
+		}
+	}
+	return "", "", "", fmt.Errorf("Couldn't parse group, version, resource from Operator %q\n", csv.Name)
```
remove capitalization

cdrage commented on pkg/service/service.go:
```diff
@@ -352,6 +389,42 @@ func getGVKFromCR(cr *olm.CRDDescription) (group, version, kind string, err erro
+	if ok {
+		err = json.Unmarshal([]byte(val), &almExamples)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to unmarshal alm-examples")
```
this ones good :)

cdrage commented on pkg/service/service.go:
```diff
@@ -352,6 +389,42 @@ func getGVKFromCR(cr *olm.CRDDescription) (group, version, kind string, err erro
+	} else {
+		// There's no alm examples in the CSV's definition
+		return nil,
+			fmt.Errorf("Could not find alm-examples in %q Operator's definition.", cr)
```
remove capitalization

cdrage commented on pkg/service/service.go:
```diff
@@ -352,6 +389,42 @@ func getGVKFromCR(cr *olm.CRDDescription) (group, version, kind string, err erro
+			return example, nil
+		}
+	}
+	return nil, errors.Errorf("Could not find example yaml definition for %q service in %q Operator's definition.\n", crd, operator)
```
remove capitalization

cdrage replied:
remove `\n`

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,76 +274,105 @@ func (o *ServiceCreateOptions) Validate() (err error) {
+				// First check if service with provided name already exists
+				svcFullName := strings.Join([]string{o.CustomResource, o.ServiceName}, "/")
+				exists, err := svc.OperatorSvcExists(o.KClient, svcFullName)
+				if kerrors.IsNotFound(err) {
```
I think this should be the opposite. If the error is a `not found` error, we should continue with exists as false.

mik-dass replied:
I checked again and the opposite condition doesn't work. We can return just false and no error from OperatorSvcExists if the resource is not found. I think that is sufficient to let the caller know about the existence. In that case we won't need the kerrors package too.

mik-dass commented on pkg/odo/cli/service/create.go:
```diff
@@ -237,76 +274,105 @@ func (o *ServiceCreateOptions) Validate() (err error) {
+			if o.ServiceName != "" && !o.DryRun {
+				// First check if service with provided name already exists
+				svcFullName := strings.Join([]string{o.CustomResource, o.ServiceName}, "/")
+				exists, err := svc.OperatorSvcExists(o.KClient, svcFullName)
```
same as above

dharmit commented:
The force push was required because of #3652 which was fixed by #3627. No other changes since the PR was lgtm'd.

mik-dass commented:
/lgtm
